fortuna/model/utils/random_features.py

```diff
@@ -0,0 +1,451 @@
+        )
+        # pylint:enable=invalid-name,not-a-mapping
+
+    def __call__(
+        self,
+        inputs: Array,
+        return_full_covmat: bool = False,
+    ) -> Tuple[Array, Array]:
+        """
+        Computes Gaussian process outputs.
+
+        Parameters
+        ----------
+        inputs: Array
+            The nd-array of shape (batch_size, ..., input_dim).
+        return_full_covmat: bool
+            Whether to return the full covariance matrix, shape
+            (batch_size, batch_size), or only return the predictive variances with
+            shape (batch_size, ).
+
+        Returns
+        -------
+        Tuple[Array, Array]
+          A tuple of predictive logits, predictive covmat and (optionally)
+          random Fourier features.
+        """
+        gp_inputs = self.norm_layer(inputs) if self.normalize_input else inputs
+        gp_features = self.random_features_layer(gp_inputs)
+
+        gp_logits = self.output_layer(gp_features)
+        gp_covmat = self.covmat_layer(
+            gp_features, gp_logits, diagonal_only=not return_full_covmat
+        )
+        return gp_logits, gp_covmat
+
+
+class RandomFourierFeatures(nn.Module):
+    """
+    A random fourier feature (RFF) layer that approximates a kernel model.
+
+    The random feature transformation is a one-hidden-layer network with
+    non-trainable weights (see, e.g., Algorithm 1 of
+    `Random Features for Large-Scale
+    Kernel Machines <https://people.eecs.berkeley.edu/~brecht/papers/07.rah.rec.nips.pdf>`_):
+
+    .. math::
+        f(x) = \gamma * cos(\mathbf{W}\mathbf{x} + \mathbf{b})
+
+    where :math:`\mathbf{W}` is the kernel matrix, :math:`\mathbf{b}` is the bias
+    and :math:`\gamma` is the output scale.
+    The forward pass logic closely follows that of the `nn.Dense` layer.
+
+    Attributes
+    ----------
+    features: int
+        The number of output units.
+    feature_scale: Optional[float]
+        Scale to apply to the output.
+        When using GP layer as the output layer of a nerual network, it is recommended to set this to 1.
+        to prevent it from changing the learning rate to the hidden layers.
+    kernel_init:  Callable[[PRNGKeyArray, Shape, Type], Array]
+         Callable[[PRNGKeyArray, Shape, Type], Array] function for the weight matrix.
+    bias_init:  Callable[[PRNGKeyArray, Shape, Type], Array]
+         Callable[[PRNGKeyArray, Shape, Type], Array] function for the bias.
+    seed: int
+        Random seed for generating random features. This will override the external RNGs.
+    dtype: Type
+        The dtype of the computation.
+    """
+    features: int
+    kernel_scale: Optional[float] = 1.0
+    feature_scale: Optional[float] = 1.0
+    kernel_init:  Callable[[PRNGKeyArray, Shape, Type], Array] = default_rbf_kernel_init
+    bias_init:  Callable[[PRNGKeyArray, Shape, Type], Array] = default_rbf_bias_init
+    seed: int = 0
+    dtype: Type = jnp.float32
```

Comment:
Will those pose any problems when we compute Cholesky factors?

Reply:
This is a great point. I do not know if this will work as expected in half precision for example. How do you suggest to proceed? Should we force the execution of this part in `jnp.float32`?